ogr/ogrsf_frmts/xodr/ogr_xodr.h

```diff
@@ -0,0 +1,261 @@
+     * Only applicable for layer types derived from meshes.
+    */
+    // TODO For lower memory consumption maybe better pass xodrRoadElements by reference?
+    OGRXODRLayer(RoadElements xodrRoadElements, std::string proj4Defn);
+    OGRXODRLayer(RoadElements xodrRoadElements, std::string proj4Defn,
+                 bool dissolveTriangulatedSurface);
+    ~OGRXODRLayer();
+};
+
+class OGRXODRLayerReferenceLine : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "ReferenceLine";
+
+    OGRXODRLayerReferenceLine(RoadElements xodrRoadElements,
+                              std::string proj4Defn);
+};
+
+class OGRXODRLayerLaneBorder : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "LaneBorder";
+
+    OGRXODRLayerLaneBorder(RoadElements xodrRoadElements,
+                           std::string proj4Defn);
+};
+
+class OGRXODRLayerRoadMark : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "RoadMark";
+
+    OGRXODRLayerRoadMark(RoadElements xodrRoadElements, std::string proj4Defn,
+                         bool dissolveTriangulatedSurface);
+};
+
+class OGRXODRLayerRoadObject : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "RoadObject";
+
+    OGRXODRLayerRoadObject(RoadElements xodrRoadElements,
+                           std::string proj4Defn);
+};
+
+class OGRXODRLayerRoadSignal : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "RoadSignal";
+
+    OGRXODRLayerRoadSignal(RoadElements xodrRoadElements, std::string proj4Defn,
+                           bool dissolveTriangulatedSurface);
+};
+
+class OGRXODRLayerLane : public OGRXODRLayer
+{
+  protected:
+    virtual void defineFeatureClass() override;
+    virtual OGRFeature *GetNextFeature() override;
+
+  public:
+    const std::string FEATURE_CLASS_NAME = "Lane";
+
+    OGRXODRLayerLane(RoadElements xodrRoadElements, std::string proj4Defn,
+                     bool dissolveTriangulatedSurface);
+};
+
+/*--------------------------------------------------------------------*/
+/*--------------------  Data source declarations ----------------------*/
+/*--------------------------------------------------------------------*/
+
+class OGRXODRDataSource : public GDALDataset
+{
+  private:
+    OGRXODRLayer **layers;
+    int nLayers;
```

Comment:
Most recent GDAL drivers uses C++11 and beyond capabilities to avoid manual memory management as much as possible:
```suggestion
    std::vector<std::unique_ptr<OGRXODRLayer>> m_apoLayers{};
```